pkg/upgrade/1.0.0-1.1.0/v1alpha1/jiva_upgrade.go

```diff
@@ -0,0 +1,408 @@
+	} else {
+		fmt.Printf("replica deployment already in %s version\n", upgradeVersion)
+	}
+	return nil
+}
+
+func patchController(controllerObj *controllerDetails, namespace string) error {
+	if controllerObj.version == currentVersion {
+		tmpl, err := template.New("controllerPatch").Parse(targetPatchTemplate)
+		if err != nil {
+			return err
+		}
+		err = tmpl.Execute(&buffer, controllerObj.patchDetails)
+		if err != nil {
+			return err
+		}
+		controllerPatch := buffer.String()
+
+		err = patchDelpoyment(
+			controllerObj.name,
+			namespace,
+			types.StrategicMergePatchType,
+			[]byte(controllerPatch),
+		)
+		if err != nil {
+			return err
+		}
+		fmt.Println(controllerObj.name, " patched")
+	} else {
+		fmt.Printf("controller deployment already in %s version\n", upgradeVersion)
+	}
+	return nil
+}
+
+func jivaUpgrade(pvName, openebsNamespace string) error {
+
+	var (
+		pvLabel         = "openebs.io/persistent-volume=" + pvName
+		replicaLabel    = "openebs.io/replica=jiva-replica," + pvLabel
+		controllerLabel = "openebs.io/controller=jiva-controller," + pvLabel
+		ns              string
+		err             error
+	)
+
+	pvObj, err := pvClient.Get(pvName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// verifying whether the pvc is deployed with DeployInOpenebsNamespace cas config
+	deployInOpenebs, err := deployClient.WithNamespace(openebsNamespace).List(
+		&metav1.ListOptions{
+			LabelSelector: pvLabel,
+		})
+	if err != nil {
+		return err
+	}
+	// check whether pvc pods are openebs namespace or not
+	if len(deployInOpenebs.Items) > 0 {
+		ns = openebsNamespace
+	} else {
+		// if pvc pods are not in openebs namespace take the namespace of pvc
+		if pvObj.Spec.ClaimRef.Namespace == "" {
+			return errors.Errorf("namespace missing for pv %s", pvName)
+		}
+		ns = pvObj.Spec.ClaimRef.Namespace
+	}
+
+	// fetching replica deployment details
+	replicaObj, err := getReplica(replicaLabel, ns)
+	if err != nil {
+		return err
+	}
+	replicaObj.patchDetails.PVName = pvName
+
+	// fetching controller deployment details
+	controllerObj, err := getController(controllerLabel, ns)
+	if err != nil {
+		return err
+	}
+	// fetching controller service details
+	controllerServiceList, err := serviceClient.WithNamespace(ns).List(
+		metav1.ListOptions{
+			LabelSelector: pvLabel,
+		})
+	if err != nil {
+		return err
+	}
+	// controllerServiceObj := controllerServiceList.Items[0]
+	controllerServiceName := controllerServiceList.Items[0].Name
+	controllerServiceVersion := controllerServiceList.Items[0].
+		Labels["openebs.io/version"]
+	if controllerServiceVersion != currentVersion &&
+		controllerServiceVersion != upgradeVersion {
+		return errors.Errorf(
+			"controller service version %s is neither %s nor %s\n",
+			controllerServiceVersion,
+			currentVersion,
+			upgradeVersion,
+		)
+	}
+
+	// replica patch
+	err = patchReplica(replicaObj, ns)
+	if err != nil {
+		return err
+	}
+	buffer.Reset()
```

Comment:
nit: Since buffer is global variable, it's good to Reset it in the same function where it's been called. It's creating confusion here.
or you can create this variable locally in the function where you are using it.